src/protocols/secure_channel/StatusReport.h

```diff
@@ -0,0 +1,103 @@
+
+    /**
+     *  Construct a StatusReport with no additional ProtocolData.
+     *
+     *  @param generalCode Required, one of the \c GeneralStatusCode values listed in \c secure_channel/Constants.h
+     *  @param protocolId Must specify a ProtocolId which consists of Vendor Id (upper 16 bits) and ProtocolId (lower 16 bits)
+     *  @param protocolCode A code defined by the specified protocol which provides more information about the status
+     */
+    StatusReport(uint16_t generalCode, uint32_t protocolId, uint16_t protocolCode);
+
+    //
+    /**
+     *  Construct a StatusReport with additional ProtocolData.
+     *
+     *  WARNING: The caller MUST ensure that the data referenced by protocolData not be freed before calling \c WriteToBuffer()
+     *      This function will not make a copy of the data.
+     *
+     *  @param generalCode Must specify a GeneralCode (see \c GeneralStatusCode )
+     *  @param protocolId Must specify a ProtocolId which consists of Vendor Id (upper 16 bits) and ProtocolId (lower 16 bits)
+     *  @param protocolCode A code defined by the specified protocol which provides more information about the status
+     *  @param protocolData A pointer to additional (and optional) protocol-specific data
+     *  @param dataLen The length of the additional (and optional) data
+     */
+    StatusReport(uint16_t generalCode, uint32_t protocolId, uint16_t protocolCode, uint8_t * protocolData, uint16_t dataLen);
+
+    /**
+     *  Read the contents of a \c PacketBuffer containing a StatusReport message and store the field values in this object.
+     *
+     *  @note If there is additional data after the Protocol Code field in the message, it is assumed to be protocol-specific data.
+     *
+     *  @note This method assumes that the Header of the message has already been consumed, and that \c PacketBuffer::Start() points
+     * to the beginning of the StatusReport data.
+     *
+     *  @param[in] buf A \c PacketBufferHandle containing the StatusReport message. This method will take ownership.
+     *
+     *  @return CHIP_ERROR Return an error if the message is malformed or buf is \c NULL
+     */
+    CHIP_ERROR Parse(System::PacketBufferHandle buf);
+
+    /**
+     *  Write the StatusReport contents into a buffer using a \c BufferWriter
+     *
+     *  @param[out] buf A \c BufferWriter which contains the buffer that will store the message fields.
+     *
+     *  @return BufferWriter Return a reference to the \c BufferWriter
+     */
+    Encoding::LittleEndian::BufferWriter & WriteToBuffer(Encoding::LittleEndian::BufferWriter & buf) const;
+
+    /**
+     *  Returns the minimum size of the buffer needed to write the message.
+     */
+    size_t Size() const;
+
+    uint16_t GetGeneralCode() const { return mGeneralCode; }
+    uint32_t GetProtocolId() const { return mProtocolId; }
+    uint16_t GetProtocolCode() const { return mProtocolCode; }
+    void GetProtocolData(uint8_t ** data, uint16_t & length) const
```

Comment:
Might be better to just have this return `ByteSpan`.

Reply:
Tennessee suggested using a `PacketBufferHandle` to represent protocolData in both the method arguments and the class member. Can you re-review the new code and see if you agree with this representation?